tests/periph_pm: make usage more intuitive and move shell commands to sys/

benemorius opened the pull request:
### Contribution description
Using tests/periph_pm can be unintuitive if you aren't already very familiar with the PM subsystem. I identified and addressed two specific causes and I think this greatly improves the situation.
1. there is no way to see which modes are currently blocked or unblocked
1. unblocking an already-unblocked mode triggers an assertion failure and kernel panic which can make the user/tester/developer think there's some serious problem when really there is no problem

This PR addresses 1) by adding a new command `pm show` to view blockers and 2) by adding a check to print a friendly error message when unblocking an unblocked mode.

### Testing procedure
```shell
> pm show
mode 0 blockers: 1 
mode 1 blockers: 0 
mode 2 blockers: 0 
mode 3 blockers: 0 
> unblock 2
Mode 2 is already unblocked.
1 >
```


### Issues/PRs references
I think some confusion earlier in #7897 can be avoided with these changes.

### Updates
1. Print current PM blockers on startup (execute `pm show`). In case we boot into an unresponsive shell, we at least get some direction for troubleshooting, as this tells us which power mode must have been entered.
1. Increased `pm_set()` verbosity. It may not be obvious to the user that running shell command `set %i` (which calls `pm_set(i)`) is *forcing* the CPU into the specified power mode, and that such manual forcing is not necessarily an orthodox thing to do. I tried to increase the verbosity somewhat so the process is clearer. Also, this verbosity provides an event (some printf output)  which indicates the moment when the CPU woke up, whereas previously we didn't know when it had woken up except by typing into the shell to check for a response. In some cases, the CPU might always wake immediately (or just sooner than expected) after `set %i`. This should reveal those cases.

aabadie commented:
Nice contribution! I like the idea of having access to the pm blocker status from the shell.

The new `pm` command made me think to an alternative: how about having only one `pm` command in the shell with subcommands for each actions (`reboote`, `set`, `block`, `unblock`, etc) and add a new `status` (or `show`) to print the current pm blocker status ? What do you think ?

benemorius commented:
Thanks. Yes I also think it seems like a good improvement to merge the commands into one. I only didn't for brevity.

I have even thought that these commands would be useful to move outside of `tests/periph_pm`. For instance I was testing power modes while running various `examples/` and I repeatedly tried to use these commands by accident.

I've been postponing most of the shell command changes I've wanted to make because I'd like to benefit effort-wise from (the concept of) #9538, but since we only need simple parsing here I don't mind so much.

However, I suppose this topic is at risk of rapidly deviating from what I hoped might be a relatively quick merge (ultimately I am trying to push #7897 forward along with subsequent PM efforts). I would be content to aim for a minimal PR for now, provided that consideration is given to forward compatibility with future improvements. Probably I will be looking at PM on EFM32 and EFR32 after this, so I will still have plenty of opportunity and context to make improvements here during that.

benemorius commented:
@benemorius 
> provided that consideration is given to forward compatibility with future improvements

For example the current command `pm` is forward-compatible equivalent with a future `pm show` command because `pm show` could be the default behavior when a future `pm` command is called with no arguments.

aabadie commented:
>  I would be content to aim for a minimal PR for now, provided that consideration is given to forward compatibility with future improvements.

Having all subcommands under `pm` was just a suggestion. It's great if you had the same idea. I'm fine doing this in follow-up PRs.

> the current command pm is forward-compatible equivalent with a future pm show command because pm show could be the default behavior when a future pm command is called with no arguments

Calling `pm` alone should print the basic usage of the command with available subcommands. So it won't be forward-compatible in this case.

Maybe you could rename the new `pm` command in `status` (or `show` as you prefer) in this PR. This will smooth the future merge in a single command.

benemorius commented:
I changed the command to `pm show` and moved it out of `tests/periph_pm` into `sys/shell/commands` so it can be used also to test that power modes behave correctly in various examples.

Also the command is run automatically when `tests/periph_pm` starts in case the shell isn't responsive:
```
main(): This is RIOT! (Version: 2019.07-devel-929-g881951-HEAD)
This application allows you to test the CPU power management.
The available power modes are 0 - 3. Lower-numbered power modes
save more power, but may require an event/interrupt to wake up
the CPU. Reset the CPU if needed.
mode 0 blockers: 0 
mode 1 blockers: 0 
mode 2 blockers: 0 
mode 3 blockers: 0 
lowest allowed mode: 0
>
```

roberthartung commented:
Will have a look tomorrow :)

roberthartung commented:
We should think about how this behaves if we have a low-power uart on a board that can wakeup the system. I know that there was a PR/Issue for that #7947 and #11310 - will have to look what they do :)

benemorius commented:
> We should think about how this behaves if we have a low-power uart on a board that can wakeup the system

In that case `pm show` will show the blocker for that and you can verify the expected behavior by manually unblocking it and observing that uart rx stops working. Also depending on the uart driver it may be possible that you can type a character at just the right time (while `pm_blocker` is being read) to see a blocker that is only active during frame reception.

Also if the `uart_write()` implementation is nonblocking, it's possible for `pm_blocker` to be sampled while the uart is actively transmitting or a DMA transfer is in progress and then those blockers will be shown. I don't think there are any uart drivers like that in master yet, but I've observed this case in a local branch with DMA buffered transmit and it works as expected.

roberthartung commented:
I will give this another test so we can merge asap!

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    if (!strcmp(argv[1], "block"))
+    {
```
The opening brace should be on the same line as `if`. Same for the `unblock` and `set` subcommands.

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+{
+    if (!strcmp(argv[1], "show")) {
+        if (argc != 2) {
+            printf("usage: pm show: display current blockers for each power mode\n");
```
```suggestion
            puts("usage: pm show: display current blockers for each power mode");
```

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+extern volatile pm_blocker_t pm_blocker; /* sys/pm_layered/pm.c */
+
+static void _print_usage(void) {
+    printf("Usage:\n");
```
```suggestion
    puts("Usage:");
```

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+
+static void _print_usage(void) {
+    printf("Usage:\n");
+    printf("\tpm show: display current blockers for each power mode\n");
```
```suggestion
    puts("\tpm show: display current blockers for each power mode");
```

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+static void _print_usage(void) {
+    printf("Usage:\n");
+    printf("\tpm show: display current blockers for each power mode\n");
+    printf("\tpm set <mode>: manually set power mode (lasts until WFI returns)\n");
```
```suggestion
    puts("\tpm set <mode>: manually set power mode (lasts until WFI returns)");
```

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+    printf("Usage:\n");
+    printf("\tpm show: display current blockers for each power mode\n");
+    printf("\tpm set <mode>: manually set power mode (lasts until WFI returns)\n");
+    printf("\tpm block <mode>: manually block power mode\n");
```
```suggestion
    puts("\tpm block <mode>: manually block power mode");
```

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+    printf("\tpm show: display current blockers for each power mode\n");
+    printf("\tpm set <mode>: manually set power mode (lasts until WFI returns)\n");
+    printf("\tpm block <mode>: manually block power mode\n");
+    printf("\tpm unblock <mode>: manually unblock power mode\n");
```
```suggestion
    puts("\tpm unblock <mode>: manually unblock power mode");
```

aabadie commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    printf("CPU is entering power mode %d.\n", mode);
+    printf("Now waiting for a wakeup event...\n");
```
```suggestion
    puts("Now waiting for a wakeup event...");
```

aabadie commented on tests/periph_pm/main.c:
```diff
@@ -28,10 +28,20 @@
 
 #ifdef MODULE_PM_LAYERED
+
+/* TODO deduplicate this definition with the one in sys/pm_layered/pm.c */
```
You could move the definition to `pm_layered.h`. Is there a reason that prevents it ?

roberthartung replied:
I don't think so. @kaspar030 what do you think? Could we integrate this into #13380 as well?

aabadie commented on tests/periph_pm/main.c:
```diff
@@ -145,10 +155,14 @@ static int cmd_set(int argc, char **argv)
 
-    printf("CPU will enter power mode %d.\n", mode);
+    printf("CPU is entering power mode %d.\n", mode);
+    printf("Now waiting for a wakeup event...\n");
```
```suggestion
    puts("Now waiting for a wakeup event...");
```

benemorius commented:
Sorry for the delay. I moved pm_blocker_t to pm_layered.h

benemorius commented:
I think I've addressed all your comments @aabadie

gschorcht commented on sys/shell/commands/sc_pm.c:
```diff
@@ -0,0 +1,166 @@
+    return 0;
+}
+
+int _pm(int argc, char **argv)
```
I would suggest a name that is a bit longer to avoid future naming conflicts. Most command handlers are calles something like `_pm_handler`

benemorius commented:
@gschorcht 
1) Oops I knew something mundane was missing from this. I hadn't bothered to deduplicate that yet. I added a fixup.

2) Ok I added `pm off`

3) I think you addressed this in #13679 right? Would you be content to address this part in a followup PR?

gschorcht commented:
> > 3\. Indeed, a question is whether the `pm` command should be included whenever the feature `periph_pm` is provided or the module `pm_layered` is used. It depends on the purpose. If it is to be a standard command, it should be included whenever the feature `periph_pm` is provided or module `pm_layered` is used. However, if it is only intended for testing, it should only be included explicitly. In any case, it will increase the the ROM size.

> 3\. I think you addressed this in #13679 right?

Yes.

> Would you be content to address this part in a followup PR?

I would like to have it considered in this PR for the following reasons:

1. The commands strongly affect the system state and in the worst case can lead to an unusable system. Therefore, an application developer might not want to expose these commands in his application.
2. The commands are pulled in automatically with module `shell_commands` when feature `pm_layered` is provided. So it increases the code size and RAM requirements even if it is not used.

Therefore, I would suggest to provide the command by a separate pseudomodule `pm_cmd` which also might be pulled in automatically if feature `pm_layered` is provided. The difference to your approach would be that it could be disabled explicitly.

@aabadie How do you think about that?

gschorcht commented:
> Maybe we should rather turn the `pm_layered` module into a `DEFAULT_MODULE` for them.
> Then it could be disabled explicitly at the application level using `DISABLE_MODULE += pm_layered`. If it makes sense, I wouldn't leave such a change for another PR.

Hm, that would be a bit like all or nothing. As I said, my biggest concern is that an application developer might not want to expose the `pm_layered` commands. Usually, it isn't an option to disable the power management. Therefore, I would separate the feature `pm_layered` and the `pm` command.

What about following approach:

- makefiles/pseudomodules.inc.mk:
   ```diff
   + PSEUDOMODULES += pm_cmd
   ```
- Makefile.dep:
   ```diff
  + ifneq (,$(filter periph_pm, $(USEMODULE)))
  +     DEFAULT_MODULE += pm_cmd
  + endif
   ```
- sys/shell/commands/shell_commands.c:
   ```diff
   - #ifdef MODULE_PM_LAYERED
   + #ifdef MODULE_PM_CMD
         extern int _pm_handler(int argc, char **argv);
     #endif
   ```
- sys/shell/commands/sc_pm.c:
  ```
   + #ifdef MODULE_PM_LAYERED
         <pm_layered> functions
     #endif
  ```
This would pull in by default `pm` command, `pm off`, it would leave `periph_pm` and `pm_layered` enabled as defined by the CPU, `pm off` would be available without `pm_layered` enabled and an application developer might use `DISABLE_MODULE += pm_cmd` to disable the commands.

aabadie commented:
> that would be a bit like all or nothing.

Indeed but it's generally how the `shell_commands` module works. If a module is loaded and provides shell commands, then pulling the shell + shell_commands modules automatically add the corresponding commands. This is the case for `ps` and a lot of other modules.
We could make this more configurable but I think this is out of the scope of this PR.

What you propose might then be extended to a lot more things then just only pm (by the way, how about calling the pseudomodule shell_commands_pm instead of pm_cmd ?)

aabadie commented:
> If it makes sense, I wouldn't leave such a change for another PR.

Errata: in https://github.com/RIOT-OS/RIOT/pull/11731#pullrequestreview-384527618, I wanted to say the opposite, e.g. _If it makes sense, I **would** leave such a change for another PR_

gschorcht commented:
> > that would be a bit like all or nothing.
> 
> Indeed but it's generally how the `shell_commands` module works. If a module is loaded and provides shell commands, then pulling the shell + shell_commands modules automatically add the corresponding commands. This is the case for `ps` and a lot of other modules.

Agreed. But, some shell commands should be available always such as `ps` and `reboot`. And these commands are not a threat to the system either, i.e. the user can't put the system in an unreachable state. With 'pm set 1' he can.

> We could make this more configurable but I think this is out of the scope of this PR.

Agreed.

> (by the way, how about calling the pseudomodule `shell_commands_pm` instead of `pm_cmd` ?)

I would be fine with such a name, it's just a 'bit' longer :wink:

BTW, @benpicco was the first who opened the discussion in PR https://github.com/RIOT-OS/RIOT/pull/13638#discussion_r393867492 :dark_sunglasses: It's exactly the same here.

gschorcht commented:
> > If it makes sense, I wouldn't leave such a change for another PR.
> 
> Errata: in [#11731 (review)](https://github.com/RIOT-OS/RIOT/pull/11731#pullrequestreview-384527618), I wanted to say the opposite, e.g. _If it makes sense, I **would** leave such a change for another PR_

IMHO, if there is something that would make sense, then it is better to do it right away. Once it is merged, it is out of focus and will probably never be done :worried:

gschorcht commented:
> > that would be a bit like all or nothing.
> 
> Indeed but it's generally how the `shell_commands` module works. If a module is loaded and provides shell commands, then pulling the shell + shell_commands modules automatically add the corresponding commands. This is the case for `ps` and a lot of other modules.
> We could make this more configurable but I think this is out of the scope of this PR.
> 
> What you propose might then be extended to a lot more things then just only `pm` (by the way, how about calling the pseudomodule `shell_commands_pm` instead of `pm_cmd`?)

@aabadie How do we continue with this PR? I really like that feature. That's why I provided a PR some days a go :wink: So I'm not strictly against merging the PR as is. On the other hand, my proposed change could be realized quite fast.

 @benpicco What is your opinion? You started the discussion not to pull in commands with features by default that are mainly for debugging purposes.

aabadie commented:
I think we should follow the same design that used with other shell_commands and then keep this PR in its current state.

Having something more configurable with sub-shell_commands could be done later and you seem to be interested by this. I can help for reviewing.

gschorcht commented:
> I think we should follow the same design that used with other shell_commands and then keep this PR in its current state.

OK. Agreed. But a last question remains. At the moment, `pm` command is pulled in only when `pm_layered` is used. I'm wondering whether it would be better to pull it in when `periph_pm` is enabled and to embed `pm_layered` related commands in `#ifdef MODULE_PM_LAYERED ... #endif` so that at least the `off` command is available without `pm_layered`

aabadie commented:
> I'm wondering whether it would be better to pull it in when periph_pm is enabled and to embed pm_layered related commands in #ifdef MODULE_PM_LAYERED ... #endif so that at least the off command is available without pm_layered

I think it's a good idea.

So basically use:

- in `sys/shell/commands/Makefile`
```mk
ifneq (,$(filter periph_pm,$(USEMODULE)))
  SRC += sc_pm.c
endif
```

- in `sys/shell/commands/sc_pm.c`
```
#ifdef MODULE_PM_LAYERED
#include "pm_layered.h"

extern volatile pm_blocker_t pm_blocker; /* sys/pm_layered/pm.c */
#endif /* MODULE_PM_LAYERED */


static void _print_usage(void) {
    puts("Usage:");
#ifdef MODULE_PM_LAYERED
    puts("\tpm show: display current blockers for each power mode");
    puts("\tpm set <mode>: manually set power mode (lasts until WFI returns)");
    puts("\tpm block <mode>: manually block power mode");
    puts("\tpm unblock <mode>: manually unblock power mode");
#endif /* MODULE_PM_LAYERED */
    puts("\tpm off: call pm_off()");
}

#ifdef MODULE_PM_LAYERED
static int check_mode(int argc, char **argv)
{
[...]

#endif /* MODULE_PM_LAYERED */

static int cmd_off(char *arg)
{
    (void)arg;

    pm_off();

    return 0;
}

[...] etc

```

@benemorius, would mind updating this PR once more ? After that we should be good to go. Thanks!

aabadie commented:
@benemorius @gschorcht, let's move forward with this one. Is it ok for both of you if I take over this PR and open a new one adding the requested changes ?

benpicco commented:
@aabadie I think you mean @benemorius :wink:

aabadie commented:
> I think you meant @benemorius

Yes, and I don't see where is the typo :)

aabadie commented:
I finally went for the force-push option in your branch @benemorius.
After addressing the current comments, I found other minor things and now it should be good.
I also tested this PR locally and it works as expected.

Maybe someone could give a final round ? @fjmolinas ?

aabadie commented:
Some boards doesn't have enough memory now. I'll exclude them from the CI.

aabadie commented:
And go!
